src/Xamarin.Android.Build.Tasks/Tasks/ResolveSdksTask.cs

```diff
@@ -81,6 +87,11 @@ public override bool RunTask ()
 			MonoAndroidLibPath  = MonoAndroidHelper.GetOSLibPath () + Path.DirectorySeparatorChar;
 			AndroidBinUtilsPath = MonoAndroidBinPath + "ndk" + Path.DirectorySeparatorChar;
 
+			var minVersion      = Version.Parse (MinimumSupportedJavaVersion);
+			var maxVersion      = Version.Parse (LatestSupportedJavaVersion);
+
+			JavaSdkPath         = MonoAndroidHelper.GetJdkInfo (this.CreateTaskLogger (), JavaSdkPath, minVersion, maxVersion)?.HomePath;
```

Comment:
This task gets called per Android project on every build. So I could see this adding incremental build time, especially in a larger solution. I think I've seen the `JdkInfo` lookup take 100ms before.

Is there a way we could pass the min/max JDK version through here instead:

https://github.com/xamarin/xamarin-android/blob/57e1a37c443c90f3487a93a312fbfef3441d2769/src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs#L116

If we want something easier to backport, maybe hardcoding something in xamarin-android-tools would work, too?

Reply:
Here is where I saw JDK lookups taking time: https://github.com/xamarin/xamarin-android/commit/e624434344ed54f39ed174d4c3a46a6cb06a7302

Reply:
1. Should we *merge* `<ResolveJdkJvmPath/>` and `<ResolveSdks/>`, "reverting" *and also* adding the `BuildEngine4.RegisterTaskObject()` invocations to `<ResolveSdks/>`?
2. I saw e624434 in passing, and I *wonder* what exactly was being measured?  (Aside: *why* does `JdkInfo` take 47ms to construct?!  Is it the `Lazy<T>` construction, used to parse `-XshowSettings:properties -version` output?)

Not that I'm saying you didn't see that improvement.  I'm wondering about the *environment* of the measurement.  In particular, was `msbuild /p:JavaSdkDirectory=…` used to capture that timing?

If `$(JavaSdkDirectory)` *isn't* set -- which I assume is the default and more common occurrence -- then we'll almost certainly "always" hit https://github.com/xamarin/xamarin-android-tools/blob/683f37508b56c76c24b3287a5687743438625341/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L69

```csharp
JavaSdkPath    = GetValidPath (ValidateJavaSdkLocation,     javaSdkPath,    () => PreferedJavaSdkPath,      () => GetJavaSdkPaths ());
```

`javaSdkPath` will be null, necessitating that we first try `() => PreferedJavaSdkPath`.

For most customer machines, this should return a valid value.  This (presumably) is a net-win, and would be a reason to look into updating `AndroidSdkInfo` to have version constraints.

However, if `() => PreferedJavaSdkPath` fails -- which presumably is the case on this CI machine -- then we hit the `() => GetJavaSdkPaths()` codepath, which is likely why we're finding and attempting to use an unsupported JDK version.

---

Which I suppose leaves two questions:

 1. Would `BuildEngine4.RegisterTaskObject()` within `<ResolveSdks/>` fix the problem?
 2. Do we know what about `JdkJvmPath` lookup is slow?  The directory traversal?

In retrospect, the "odd" thing about the pre-e624434 world order is that `ResolveSdks.GetJvmPath()` even existed in the first place, as constructed; once we have a JDK path, why would `GetJvmPath()` need to call `JdkInfo.GetKnownSystemJdkInfos()` again?

Looks like that was always the case, though:

https://github.com/xamarin/xamarin-android/commit/1a2eb953af76b76b2443b744894523ddfe239ecf#diff-ec49e6de47ef6cb5b59095d242a56acce5830d39402bf075722fce7dd1326612R105

Also odd that I kinda/sorta mentioned the "oddness" of having a `JdkInfo.GetKnownSystemJdkInfos()` invocation, then forgot about how odd that was: https://github.com/xamarin/xamarin-android/pull/2153#discussion_r215961078

I don't see why, given a valid JDK path, `JdkJvmPath` lookup should be *so slow*; via `csharp`:

```csharp
using Xamarin.Android.Tools;
using System.Diagnostics;
using System.IO
var validJdkPath=Path.Combine(Environment.GetEnvironmentVariable("HOME"), "Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25");
var s = Stopwatch.StartNew(); var p = new JdkInfo(validJdkPath).JdkJvmPath; s.Stop();
s.ElapsedMilliseconds
// 26 on first invocation, 0 on subsequent iterations.  Behold the JIT!
```

So maybe `JdkJvmPath` *is* slow, *because* of JIT overheads (ugh), and no amount of caching can fix things when we're concerned about design-time builds on the *first* build.  It doesn't matter if subsequent builds are faster (yay JIT!), when we care about the Design-Time build, which won't need `JdkJvmPath`…

But what about the JDK itself?

```csharp
using Xamarin.Android.Tools;
using System.Diagnostics;
using System.IO
var s = Stopwatch.StartNew(); var p = new AndroidSdkInfo().JavaSdkPath; s.Stop();
s.ElapsedMilliseconds;
print(p);
```

This is "differently fun"; if I edit `$HOME/.config/xbuild/monodroid-config.xml` and *remove* the `<java-sdk/>` element -- removing the "preferred JDK" location -- then the above takes 81ms to run, while if I have a `<java-sdk/>` element -- providing a "preferred JDK" location -- then the above takes 55ms to run, a savings of 26ms when the preferred JDK location is set and exists.

Reply:
When I was testing https://github.com/xamarin/xamarin-android/commit/e624434344ed54f39ed174d4c3a46a6cb06a7302, I probably didn't have `JavaSdkDirectory` set at all and was testing from the command-line.

Let me just test this change tomorrow on Windows/.NET framework with a Release build and see how long it takes.

Were your numbers above a `Release` build?